synapse/storage/util/id_generators.py

```diff
@@ -273,6 +273,19 @@ def _load_current_ids(
 
         # Load the current positions of all writers for the stream.
         if self._writers:
+            # We delete any stale entries in the positions table. This is
+            # important if we add back a writer after a long time; we want to
+            # consider that a "new" writer, rather than using the old stale
+            # entry here.
+            sql = """
+                DELETE FROM stream_positions
+                WHERE
+                    stream_name = ?
+                    AND instance_name != ALL(?)
```

Comment:
Confusing, `instance_name != ALL(?)` follows [this logic](https://www.postgresqltutorial.com/postgresql-all/):

> `column_name != ALL (subquery)` the expression evaluates to true if a value is not equal to any value returned by the subquery.

So this is saying to delete any stream position where the stream name matches our current stream and the instance name is not any of the current writers.